include/deal.II/base/tensor.h

```diff
@@ -1487,348 +1479,221 @@ operator * (const Tensor<rank_1, dim, Number> &src1,
-          break;
-
-        default:
-          Assert (false,
-                  (typename Tensor<2,dim,Number>::ExcInvalidTensorContractionIndex (index2)));
-        };
-      break;
-    case 2:
-      switch (index2)
-        {
-        case 1:
-          for (unsigned int i=0; i<dim; ++i)
-            for (unsigned int j=0; j<dim; ++j)
-              for (unsigned int k=0; k<dim; ++k)
-                dest[i][j] += src1[i][k] * src2[k][j];
-          break;
-        case 2:
-          for (unsigned int i=0; i<dim; ++i)
-            for (unsigned int j=0; j<dim; ++j)
-              for (unsigned int k=0; k<dim; ++k)
-                dest[i][j] += src1[i][k] * src2[j][k];
-          break;
-
-        default:
-          Assert (false,
-                  (typename Tensor<2,dim,Number>::ExcInvalidTensorContractionIndex (index2)));
-        };
-      break;
-
-    default:
-      Assert (false, (typename Tensor<2,dim,Number>::ExcInvalidTensorContractionIndex (index1)));
-    };
-}
-
-
-/**
- * Contract a tensor of rank 3 with a tensor of rank 1. The contraction is
- * performed over index <tt>index1</tt> of the first tensor.
+ * @note The position of the index is counted from 0, i.e.,
+ * $0\le\text{index_i}<\text{range_i}$.
  *
- * Note that the number of the index is counted from 1 on, not from zero as
- * usual.
+ * @note In case the contraction yields tensor of rank 0 the scalar
+ * number is returned as an unwrapped number type.
  *
  * @relates Tensor
- * @author Wolfgang Bangerth, 1998
+ * @author Matthias Maier, 2015
  */
-template <int dim, typename Number>
+template <int index_1, int index_2, int index_3, int index_4,
+          int rank_1, int rank_2, int dim,
+          typename Number, typename OtherNumber>
 inline
-void contract (Tensor<2,dim,Number>       &dest,
-               const Tensor<3,dim,Number> &src1,   const unsigned int index1,
-               const Tensor<1,dim,Number> &src2)
-{
-  dest.clear ();
-
-  switch (index1)
-    {
-    case 1:
-      for (unsigned int i=0; i<dim; ++i)
-        for (unsigned int j=0; j<dim; ++j)
-          for (unsigned int k=0; k<dim; ++k)
-            dest[i][j] += src1[k][i][j] * src2[k];
-      break;
-
-    case 2:
-      for (unsigned int i=0; i<dim; ++i)
-        for (unsigned int j=0; j<dim; ++j)
-          for (unsigned int k=0; k<dim; ++k)
-            dest[i][j] += src1[i][k][j] * src2[k];
-      break;
-
-    case 3:
-      for (unsigned int i=0; i<dim; ++i)
-        for (unsigned int j=0; j<dim; ++j)
-          for (unsigned int k=0; k<dim; ++k)
-            dest[i][j] += src1[i][j][k] * src2[k];
-      break;
-
-    default:
-      Assert (false,
-              (typename Tensor<2,dim,Number>::ExcInvalidTensorContractionIndex (index1)));
-    };
+typename Tensor<rank_1 + rank_2 - 4, dim, typename ProductType<Number, OtherNumber>::type>::tensor_type
+contract (const Tensor<rank_1, dim, Number> &src1,
+          const Tensor<rank_2, dim, OtherNumber> &src2)
```

Comment:
*mhm* I originally chose ```contract``` because I had a variadic template in mind. But the overhead to actually support it turned out to be a bit too much..

I'll rename.